source/grid/tria.cc

```diff
@@ -5177,6 +5179,1237 @@ namespace internal
+             ++line)
+          line->clear_user_flag();
+
+        for (typename Triangulation<dim, spacedim>::quad_iterator quad =
+               triangulation.begin_quad();
+             quad != triangulation.end_quad();
+             ++quad)
+          quad->clear_user_flag();
+
+        unsigned int needed_vertices     = 0;
+        unsigned int needed_lines_single = 0;
+        unsigned int needed_quads_single = 0;
+        unsigned int needed_lines_pair   = 0;
+        unsigned int needed_quads_pair   = 0;
+        for (int level = triangulation.levels.size() - 2; level >= 0; --level)
+          {
+            unsigned int new_cells = 0;
+
+            for (const auto &acell :
+                 triangulation.active_cell_iterators_on_level(level))
+              if (acell->refine_flag_set())
+                {
+                  Assert(acell->refine_flag_set() ==
+                           RefinementCase<dim>::cut_xyz,
+                         ExcInternalError());
+
+                  if (acell->reference_cell_type() == ReferenceCell::Type::Hex)
+                    {
+                      ++needed_vertices;
+                      needed_lines_single += 6;
+                      needed_quads_single += 12;
+                      new_cells += 8;
+                    }
+                  else if (acell->reference_cell_type() ==
+                           ReferenceCell::Type::Tet)
+                    {
+                      needed_lines_single += 1;
+                      needed_quads_single += 8;
+                      new_cells += 8;
+                    }
+                  else
+                    {
+                      Assert(false, ExcInternalError());
+                    }
+
+                  for (const auto face : acell->face_indices())
+                    if (acell->face(face)->number_of_children() < 4)
+                      acell->face(face)->set_user_flag();
+
+                  for (const auto line : acell->line_indices())
+                    if (acell->line(line)->has_children() == false)
+                      acell->line(line)->set_user_flag();
+                }
+
+            const unsigned int used_cells =
+              std::count(triangulation.levels[level + 1]->cells.used.begin(),
+                         triangulation.levels[level + 1]->cells.used.end(),
+                         true);
+
+            reserve_space(*triangulation.levels[level + 1],
+                          used_cells + new_cells,
+                          3,
+                          spacedim);
+
+            reserve_space(triangulation.levels[level + 1]->cells, new_cells);
+          }
+
+        for (typename Triangulation<dim, spacedim>::quad_iterator quad =
+               triangulation.begin_quad();
+             quad != triangulation.end_quad();
+             ++quad)
+          {
+            if (quad->user_flag_set() == false)
+              continue;
+
+            if (quad->reference_cell_type() == ReferenceCell::Type::Quad)
```

Comment:
we let this sit for too long (this is partly my fault) so it's gotten stale - we will need to rebase and fix the compilation errors (e.g., this is now just `reference_cell`)